R/format_data.R

```diff
@@ -996,6 +996,310 @@ fmt_percent <- function(data,
+#'   1, 2, or 3 digits. Several options provide a fixed denominator: `"/2"` (for
+#'   fractions as halves), `"/4"` (quarters), `"/8"` (eighths), `"/16"`
+#'   (sixteenths), and `"1/100"` (hundredths).
+#' @param layout For HTML output, the `"diagonal"` layout is the default. This
+#'   will generate fractions that are typeset with raised/lowered numerals and a
+#'   virgule. The `"inline"` layout places the numerals of the fraction on the
+#'   baseline and uses a standard slash character.
+#'
+#' @return An object of class `gt_tbl`.
+#'
+#' @family Format Data
+#' @section Function ID:
+#' 3-6
+#'
+#' @import rlang
+#' @export
+fmt_fraction <- function(
+    data,
+    columns,
+    rows = everything(),
+    accuracy = c("low", "med", "high", "/2", "/4", "/8", "/16", "/100"),
+    layout = c("diagonal", "inline"),
+    use_seps = TRUE,
+    pattern = "{x}",
+    sep_mark = ",",
+    dec_mark = ".",
+    locale = NULL
+) {
+
+  accuracy <- match.arg(accuracy)
+  layout <- match.arg(layout)
+
+  # Perform input object validation
+  stop_if_not_gt(data = data)
+
+  # Resolve the `locale` value here with the global locale value
+  locale <- resolve_locale(data = data, locale = locale)
+
+  # Stop function if any columns have data that is incompatible
+  # with this formatter
+  if (!column_classes_are_valid(
+    data = data,
+    columns = {{ columns }},
+    valid_classes = c("numeric", "integer")
+  )
+  ) {
+    stop(
+      "The `fmt_fraction()` function can only be used on `columns` with numeric data",
+      call. = FALSE
+    )
+  }
+
+  # In the inline layout (where there is no special typesetting of the
+  # fraction part) we must include a space to distinguish between the
+  # whole number and fraction parts
+  # if (layout == "inline") {
+  #   incl_space <- TRUE
+  # }
+
+  # Use locale-based marks if a locale ID is provided
+  sep_mark <- get_locale_sep_mark(locale, sep_mark, use_seps)
+  dec_mark <- get_locale_dec_mark(locale, dec_mark)
+
+  # Pass `data`, `columns`, `rows`, and the formatting
+  # functions as a function list to `fmt()`
+  fmt(
+    data = data,
+    columns = {{ columns }},
+    rows = {{ rows }},
+    fns = num_fmt_factory_multi(
+      pattern = pattern,
+      format_fn = function(x, context) {
+
+        # Round all values of x to 3 digits with the R-H-U method of
+        # rounding (for reproducibility purposes)
+        x <- round_gt(x, 3)
+
+        # Determine which of `x` are finite values
+        x_is_a_number <- is.finite(x)
+
+        # Divide the `x` values in 'big' and 'small' components; delay the
+        # formatting of `big_x` until it is appropriately rounded on the
+        # basis of the fractions obtained at the desired accuracy
+        big_x <- trunc(x)
+        small_x <- abs(x - big_x)
+
+        # Get the correct minus mark based on the output context
+        minus_mark <- context_minus_mark(context = context)
+
```

Comment:
Instead of having hardcoded 2/4/8/16/100 denominators, let's replace that with dynamically calculating an arbitrary denom.

```r
gcd <- function(x,y) {
  r <- x%%y;
  return(ifelse(r, gcd(y, r), y))
}

make_frac <- function(x, denom, simplify = TRUE) {
  big_x <- trunc(x)
  small_x <- x %% 1
  
  numer <- round(small_x * denom)
  
  if (simplify) {
    denom <- rep_len(denom, length(x))
    factor <- gcd(numer, denom)
    numer <- numer / factor
    denom <- denom / factor
  }
  
  ifelse(numer == denom, "1",
    ifelse(numer == 0, "0", paste0(numer, "/", denom))
  )
}
```

Reply:
And add a function param for whether the fraction should be simplified; this wouldn't apply to low/med/high, only to cases where they're asking for a specific denominator.